src/Compilers/CSharp/Portable/Symbols/Source/SourceExtensionTypeSymbol.cs

```diff
@@ -112,31 +112,27 @@ protected override void CheckBaseExtensions(BindingDiagnosticBag diagnostics)
                     }
                 }
             }
-        }
-
-        // PROTOTYPE consider caching and using a property like InterfacesAndTheirBaseInterfacesNoUseSiteDiagnostics
-        private MultiDictionary<NamedTypeSymbol, NamedTypeSymbol> GetExtensionsAndTheirBaseExtensionsNoUseSiteDiagnostics()
-        {
-            var baseExtensions = this.BaseExtensionsNoUseSiteDiagnostics;
-            var resultBuilder = new MultiDictionary<NamedTypeSymbol, NamedTypeSymbol>(baseExtensions.Length,
-                SymbolEqualityComparer.CLRSignature, SymbolEqualityComparer.ConsiderEverything);
 
-            foreach (var baseExtension in baseExtensions)
+            MultiDictionary<NamedTypeSymbol, NamedTypeSymbol> mapAllBaseExtensionsDuplicates()
             {
-                if (resultBuilder.Add(baseExtension, baseExtension))
+                var baseExtensions = this.AllBaseExtensionsNoUseSiteDiagnostics;
+                var resultBuilder = new MultiDictionary<NamedTypeSymbol, NamedTypeSymbol>(baseExtensions.Length,
+                    SymbolEqualityComparer.CLRSignature, SymbolEqualityComparer.ConsiderEverything);
+
+                foreach (var baseExtension in baseExtensions)
                 {
-                    // PROTOTYPE we need to collect all base extensions from baseExtension too
+                    resultBuilder.Add(baseExtension, baseExtension);
                 }
-            }
 
-            return resultBuilder;
+                return resultBuilder;
+            }
         }
 
         internal sealed override TypeSymbol? GetDeclaredExtensionUnderlyingType()
-            => GetDeclaredExtensionInfo().UnderlyingType;
+            => GetDeclaredExtensionInfo(basesBeingResolved: null).UnderlyingType;
```

Comment:
>basesBeingResolved: null [](http://example.com/codeflow?start=40&length=24)

This looks suspicious. We do not have ```basesBeingResolved```, but will try to bind base extensions as well. However, it looks like in order to avoid cycles while binding base extensions, we need ```basesBeingResolved```. Is there a guarantee that we will never end-up here while binding base extensions? Otherwise, we might want to split binding of the underling type and base extension, so that base extensions are not bound when someone asks for an underlying type.

Reply:
Like you, this method gave me pause and made me hesitate a bit.
I'm not 100% sure this is error-proof, but I've tried various scenarios without succeeding to break it. 
I don't think we can build a cycle via the underlying type because as soon as we see an extension type used there, we break the cycle (store no underlying type) without asking more questions.

If we do encounter a problem, we can either split the computation as you suggested, or add a `basesBeingResolved` parameter to `GetDeclaredExtensionUnderlyingType`, or figure out some other solution. I could also add a PROTOTYPE comment if you think we should pro-actively track this further.